Document VertexOnlyMesh in manual (#2912)
firedrakeproject/firedrake: docs/source/point-evaluation.rst

Firedrake can evaluate :py:class:`~.Function`\s at arbitrary physical
points.  This feature can be useful for the evaluation of the result
of a simulation, or for creating expressions which contain point evaluations.
Three APIs are offered to this feature: two Firedrake-specific ones, and one
from UFL.

> Firedrake API 1: 'at' method

Reviewer: ```suggestion
Firedrake convenience function
```

it. So if we have a vertex-only mesh :math:`\Omega_v` with :math:`N` vertices
at points :math:`\{x_i\}_{i=0}^{N-1}` and we have interpolated a function
:math:`f` onto it giving a new function :math:`f_v` then

.. math::

   \int_{\Omega_v} f_v \, dx = \sum_{i=0}^{N-1} f(x_i).

These equivalent expressions for point evaluation

.. math::

   \sum_{i=0}^{N-1} f(x_i) = \sum_{i=0}^{N-1} \int_\Omega f(x) \delta(x - x_i) \, dx

where :math:`N` is the number of points, :math:`x_i` is the :math:`i`\th point,
:math:`\Omega` is a 'parent' mesh, :math:`f` is a function on that mesh,
:math:`\delta` is a dirac delta distribition can therefore be written in
Firedrake using :func:`~.VertexOnlyMesh` and :func:`~.interpolate` as

.. literalinclude:: ../../tests/vertexonly/test_vertex_only_manual.py
   :language: python3
   :dedent:
   :lines: 48-62

Interacting with external point data
------------------------------------

Any set of points with associated data in our domain can be expressed as a
P0DG function on a :func:`~.VertexOnlyMesh`.

.. code-block:: python3

   vom = VertexOnlyMesh(parent_mesh, point_locations_from_elsewhere)
   P0DG = FunctionSpace(vom, "DG", 0)

>    y_pts = Function(P0DG).dat.data[:] = point_data_values_from_elsewhere

Reviewer: I think we now know this isn't parallel-safe. Maybe we should drop this pending the next fix....

Author: TODO: add a warning that the example uses an internal firedrake API (`Function(P0DG).dat.data[:]`) and that it doesn’t work in parallel. I can state that a parallel safe way of interacting with external data will be available soon.

:func:`~.VertexOnlyMesh` follows the ordering of the list of points it is given
at construction. In general :func:`~.VertexOnlyMesh` accepts any numpy array of
shape ``(num_points, point_dim)`` (or equivalent list) as the set of points to
create disconnected vertices at.

The operator for evaluation at the points specified can be
created by making an :py:class:`~.Interpolator` acting on a
:py:func:`~.TestFunction`

.. code-block:: python3

   u = TestFunction(V)
   Interpolator(u, P0DG)

For more on :py:class:`~.Interpolator`\s and interpolation see the
:doc:`interpolation <interpolation>` section.

Vector and tensor valued function spaces
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

When interpolating from vector or tensor valued function spaces, the P0DG
function space on the vertex-only mesh must be a
:py:func:`~.VectorFunctionSpace` or :py:func:`~.TensorFunctionSpace`
respectively. For example:

.. code-block:: python3

   V = VectorFunctionSpace(parent_mesh, "CG", 2)

or

.. code-block:: python3

   V = FunctionSpace(parent_mesh, "N1curl", 2)

each require

.. code-block:: python3

   vom = VertexOnlyMesh(parent_mesh, points)

>    P0DG_vec = VectorFunctionSpace(vom, "DG", 0)

Reviewer 2: Do you not need to provide `dim` as an argument?

Author: For `VectorFunctionSpace`? Nope. From the `VectorFunctionSpace` docs
```
    :arg dim: An optional number of degrees of freedom per function
       space node (defaults to the geometric dimension of the mesh).
```

Reviewer 2: Yes of course.

Mesh tolerance
--------------

If points are outside the mesh domain but ought to still be found a
``tolerance`` parameter can be set. The tolerance is relative to the size of
the mesh cells and is a property of the mesh itself

.. literalinclude:: ../../tests/vertexonly/test_vertex_only_manual.py
   :language: python3
   :dedent:
   :lines: 70-90

Keyword arguments
~~~~~~~~~~~~~~~~~

Alternatively we can modify the tolerance by providing it as an argument to the
vertex-only mesh. This will modify the tolerance property of the parent mesh.

.. warning::

   To avoid confusion, it is recommended that the tolerance be set for a mesh
   before any point evaluations are performed, rather than making use of these
   keyword arguments.

.. literalinclude:: ../../tests/vertexonly/test_vertex_only_manual.py
   :language: python3
   :dedent:
   :lines: 100-117

Note that since our tolerance is relative, the number of cells in a mesh
dictates the point loss behaviour close to cell edges. So the mesh
``UnitSquareMesh(5, 5, quadrilateral = True)`` will include the point
:math:`(1.1, 1.0)` by default.

If a mesh tolerance changes where we already have an immersed vertex-only mesh,

> such that the new tolerance ought to cause points to disappear, they will not.

Reviewer 2: I don't quite understand this paragraph.

Author: It's saying that changing mesh tolerance only effects point location when you create a new VOM

Author: Would it be clearer if I explicitly stated that?

Reviewer 2: I think so.

Author: Can you take a look at my proposed diff and see if it's alright?

Reviewer 2: Yep I'm happier with that. Thanks. If you commit those diffs I'll approve the PR.

> If a mesh tolerance changes where we already have an immersed vertex-only mesh,
> such that the new tolerance ought to cause points to disappear, they will not.

Author: ```suggestion
Changing mesh tolerance only affects point location after it has been changed.
To apply the new tolerance to a vertex-only mesh, a new vertex-only mesh must 
be created. Any existing immersed vertex-only mesh will have been created 
using the previous tolerance and will be unaffected by the change.
```

* Each process will get the same values.

Firedrake API 2: Interpolation onto a vertex-only mesh
------------------------------------------------------

Firedrake's second API for evaluating functions at arbitrary points,
interpolation onto a :func:`~.VertexOnlyMesh`, is designed for evaluating a
function at many points and for creating expressions which contain point
evaluations. It has been designed from the ground up to be entirely parallel

> compatible. Whilst :meth:`~.Function.at` produces a list of values, a

Reviewer 3: It would be nice to see a bit more detail on how to use it in parallel rather than referring to the docstring as you do later on. Is the warning in the docstring

> This means that if rank A has ``vertexcoords`` {X} that are not found in the mesh cells owned by rank A but are found in the mesh cells owned by rank B, **and rank B has not been supplied with those**, then they will be moved to rank B.

stil valid? In my interpretation reading the code - but correct me if I'm wrong - with `redundant=False` every process specifies unique points which get concatenated with an algatherv, and then every local process keeps its locally visible ones with some voting algorithm to ensure each point gets assignted to only one process.

Regardless of how the points are specified (redundant or not), I'm a little unclear on how the interpolated data is meant to be used in parallel. The `.dat.data` corresponds to the interpolated data of the locally visible detectors, but as a user I have no idea which ones they are and how they are related to the points I've specified and in which order.

Author: > stil valid? In my interpretation reading the code - but correct me if I'm wrong - with redundant=False every process specifies unique points which get concatenated with an algatherv, and then every local process keeps its locally visible ones with some voting algorithm to ensure each point gets assignted to only one process.

It is still valid, in fact it's a new feature as of #2773! The point is that you can specify a point on one rank and, if it's not on that rank but is on another, it'll be moved to that rank. See issue #2178 where this was first raised.

>Regardless of how the points are specified (redundant or not), I'm a little unclear on how the interpolated data is meant to be used in parallel. The .dat.data corresponds to the interpolated data of the locally visible detectors, but as a user I have no idea which ones they are and how they are related to the points I've specified and in which order.

I don't quite understand you here, what do you mean by detectors?

Reviewer: Well you can get the point coordinates from the VoM. However I think this exposes a hole in the current implementation. 

Think about the use case of interpolating from one mesh to another, then the point cloud starts with one decomposition. It necessarily gets rearranged to another parallel decomposition in order to match the parent mesh and do interpolation. However, the user needs the points back in the original configuration to use the data. 

I think we need to store the original rank and local number of every point and preserve this as we transmit the points around (these could be fields on the DMSwarm). Then we could provide a method that gives you back the points in the original decomposition. I guess we have to think about what the type that that would return would be.

Reviewer 3: Oh heh sorry, on the docstring: I read what it says here https://www.firedrakeproject.org/firedrake.html#firedrake.mesh.VertexOnlyMesh (which seems out-of-date): "then the vertexcoords {X} will be lost." but then I copied from the code, which is correct indeed.

That was exactly the application I was thinking of indeed.

Another application I can think of is interpolating onto an external grid, say a regular structured one. I guess in that situation you would typically use `redundant=True`, but you have the same issue that it'll be hard to piece the local interpolated values in each .dat.data back to a structured array. Is this something DMSWARM can do? We do give it the original global index of the points as provided.

Author: Yep looks like the website needs to be rebuilt @stephankramer . @dham I wonder if this moves us in a direction of needing VoMs which have no parent...

@stephankramer I suppose I could start saving the interpolated values as a field on the DMSWARM...

Author: @stephankramer Okay I've expanded the manual to include more information

Author: Also you made me realise there's a mistake in the VertexOnlyMesh API docs that I've now fixed 82f91141a1035e56941d1dbc774c6611742d2566

cross-mesh interpolation onto :func:`~.VertexOnlyMesh` gives Firedrake
:py:class:`~.Function`\s.

This is discussed in detail in :cite:`nixonhill2023consistent` but, briefly,

> the idea is that the :func:`~.VertexOnlyMesh` is a mesh whose wich represents a

Reviewer 2: ```suggestion
the idea is that a :func:`~.VertexOnlyMesh` is a mesh representing a
```

   parent_mesh = UnitSquareMesh(10, 10)

   V = FunctionSpace(parent_mesh, "CG", 2)

   # Create a function f on the parent mesh to point evaluate
   x, y = SpatialCoordinate(parent_mesh)
   f = Function(V).interpolate(x**2 + y**2)

   # 3 points (i.e. vertices) at which to point evaluate f
   points = [[0.1, 0.1], [0.2, 0.2], [0.3, 0.3]]

   vom = VertexOnlyMesh(parent_mesh, points)

   # P0DG is the only function space you can make on a vertex-only mesh
   P0DG = FunctionSpace(vom, "DG", 0)

   # Interpolation performs point evaluation
   f_at_points = interpolate(f, P0DG)

   print(f_at_points.dat.data)

will print ``[0.02, 0.08, 0.18]``, the values of :math:`x^2 + y^2` at the
points :math:`(0.1, 0.1)`, :math:`(0.2, 0.2)` and :math:`(0.3, 0.3)`.

Note that ``f_at_points`` is a :py:class:`~.Function` which takes
on *all* the values of ``f`` evaluated at ``points``. The cell ordering of a
:func:`~.VertexOnlyMesh` follows the ordering of the list of points it is given
at construction. In general :func:`~.VertexOnlyMesh` accepts any numpy array of
shape ``(num_points, point_dim)`` (or equivalent list) as the set of points to
create disconnected vertices at.

> The operator for evaluation at the points specified can be created can be

Reviewer 3: repeated "can be created"

Author: ```suggestion
The operator for evaluation at the points specified can be
```

* Each process will get the same values.

> Firedrake API 2: Interpolation onto a vertex-only mesh

Reviewer: ```suggestion
Primary API: Interpolation onto a vertex-only mesh
```

------------------------------------------------------

> Firedrake's second API for evaluating functions at arbitrary points,

Reviewer: ```suggestion
Firedrake's principal API for evaluating functions at arbitrary points,
```

interpolation onto a :func:`~.VertexOnlyMesh`, is designed for evaluating a

> function at many points and for creating expressions which contain point

Reviewer: ```suggestion
function at many points, or repeatedly, and for creating expressions which contain point
```

evaluations. It has been designed from the ground up to be entirely parallel
compatible. Whilst :meth:`~.Function.at` produces a list of values, a
cross-mesh interpolation onto :func:`~.VertexOnlyMesh` gives Firedrake
:py:class:`~.Function`\s.

This is discussed in detail in :cite:`nixonhill2023consistent` but, briefly,
the idea is that the :func:`~.VertexOnlyMesh` is a mesh whose wich represents a
point cloud domain. Each cell of the mesh is a vertex at a chosen location in
space. As usual for a mesh, we represent values by creating functions in
function spaces on it. The only function space that makes sense for a mesh
whose cells are vertices is the space of piecewise constant functions, also
known as the Polynomial degree 0 Discontinuous Galerkin (P0DG) space.

Our vertex-only meshes are immersed in some 'parent' mesh. We perform point
evaluation of a function :math:`f` defined in a function space
:math:`V` on the parent mesh by interpolating into the P0DG space on the
:func:`~.VertexOnlyMesh`. For example:

.. code-block:: python3

> 

Reviewer: Import the lines of code from a test: https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-literalinclude
